Start an analysis for Wilms tumor annotation (SCPCP000014)

JingxuanChen7 opened the pull request:
### Purpose/implementation Section

In this PR section, I'm trying to initialize an analysis module skeleton for Wilms tumor dataset SCPCP000014, as well as checking in scripts for pre-processing the provided `sce` objects.

#### Please link to the GitHub issue that this pull request addresses.

- Issue https://github.com/AlexsLemonade/OpenScPCA-analysis/issues/757
- Discussion https://github.com/AlexsLemonade/OpenScPCA-analysis/discussions/628


#### What is the goal of this pull request?

This PR files scripts for pre-processing provided `sce` objects using a standard Seurat workflow, which would be used in the following analysis in my next PR.


#### Briefly describe the general approach you took to achieve this goal.
A standard Seurat workflow (normalization, feature selection, PCA, clustering, dimensional reduction, etc) was applied to all 10 samples in this dataset.


#### If known, do you anticipate filing additional pull requests to complete this analysis module?
- Further analysis would be filed soon in other PRs. 
- This pre-processing part may also subject to change if I could identify better feature selection or dimension reduction strategies in splitting out the clusters.


### Results


#### What is the name of your results bucket on S3?

`s3://researcher-009160072044-us-east-2/cell-type-wilms-tumor-14/results/00_preprocessing_rds/`

#### What types of results does your code produce (e.g., table, figure)?

Intermediate RDS files containing pre-processed Seurat objects.

#### What is your summary of the results?

Results for this section contains 10 `.rdsSeurat` objects for further analysis.


### Provide directions for reviewers

_In this section, tell reviewers what kind of feedback you are looking for._
_This information will help guide their review._

#### What are the software and computational requirements needed to be able to run the code in this PR?

- Analysis could be executed on a virtual computer ([Standard-4XL](https://openscpca.readthedocs.io/en/latest/aws/lsfr/creating-vcs/)) via AWS Lightsail for Research.
- To setup the computing environment, run
```
conda config --add channels bioconda
conda config --add channels conda-forge
conda config --set channel_priority strict
cd /home/lightsail-user/git/OpenScPCA-analysis/analyses/cell-type-wilms-tumor-14
conda env create -f ./conda_envs/main.yml -y -n wilms-tumor-14-main
conda activate wilms-tumor-14-main
Rscript --vanilla ./conda_envs/main.R
```
- Also added to [README](https://github.com/JingxuanChen7/OpenScPCA-analysis/tree/JingxuanChen7/wilms14_pr/analyses/cell-type-wilms-tumor-14#software-requirements)

#### Are there particularly areas you'd like reviewers to have a close look at?
For the first PR, I would like to make sure if my way to setup the module skeleton, computing environment and documentation can meet the need.


#### Is there anything that you want to discuss further?
I have some questions for this PR:

1. I'm a bit confused by the scope/purpose of GitHub issues. Should we open a new issue for each PR, or multiple PRs could be included to resolve one issue? 
2. I'm using a conda environment to install all the R packages needed so far. Should I add my code to setup conda env in the Dockerfile, or is it OK to leave it as is? When I ran the code on virtual computer, I didn't use Docker in general.

Thanks for reviewing!



### Author checklists

_Check all those that apply._
_Note that you may find it easier to check off these items after the pull request is actually filed._


#### Analysis module and review

- [x] This analysis module [uses the analysis template and has the expected directory structure](https://openscpca.readthedocs.io/en/latest/contributing-to-analyses/analysis-modules/).
- [x] The analysis module `README.md` has been updated to reflect code changes in this pull request.
- [x] The analytical code is documented and contains comments.
- [x] Any results and/or plots this code produces have been added to your S3 bucket for review.

#### Reproducibility checklist

- [ ] Code in this pull request has been added to the GitHub Action workflow that runs this module.
- [ ] The dependencies required to run the code in this pull request have been added to the analysis module `Dockerfile`.
- [ ] If applicable, the dependencies required to run the code in this pull request have been added to the analysis module conda `environment.yml` file.
- [x] If applicable, R package dependencies required to run the code in this pull request have been added to the analysis module `renv.lock` file.

jashapiro reviewed (commented):
Hello @JingxuanChen7, thank you for your contribution!

I very much appreciate your efforts to keep the code you have well-organized and clear, and to include up-to-date documentation. It certainly made this initial review much easier!

Let me start by addressing your questions:

1. The GitHub issues are largely to allow everyone to track progress and break up work. In general we try to scope them so a single PR can resolve the issue, but it is also fine to have a larger issue that is addressed by a series of PRs, and of course that sometimes happens in the course of working an issue that you might have though would only take a single PR, but ends up requiring a bit more effort than can be reviewed in a single PR. 
2. It is fine to leave the Dockerfile as is. We do not expect you to use Docker for your own work, especially early in development when the package requirements are likely to be changing with some frequency. As your analysis gets more complete, we may update the dockerfile to match the environment you are using. 

As for the module setup itself, I very much appreciate you setting up a distinct environment for your work with all of your package dependencies included. It seems like your project is mostly based in R? If that is the case, we would in general prefer that you set up your environment with `renv`, as that allows us to more easily track the versions of specific packages and dependencies. The versions available on conda are often a bit out of date as well, so this allows us to keep up better with current versions of Bioconductor, etc.

If you would like any help setting that up, please let me know, or if you have a compelling reason that you would prefer to stick with `conda` for this purpose, we can also consider that for this project. 

Since I think it makes sense to start with the project as a whole, I did not spend too much time looking at your processing code. I did note, however, that you are performing normalization and dimension reduction within Seurat. This may be required by your downstream analysis, but I will note that we do include default normalized and dimension reduction matrices, and it may be more efficient to maintain the pre-computed values in the SingleCellExperiment object rather than recalculating them. If this is something you would like to pursue, I am happy to answer questions about where those values are located. The objects and processing steps are generally described in the portal documentation at https://scpca.readthedocs.io/en/stable/sce_file_contents.html

Again, thank you very much for your first submission! I look forward to working with you as we incorporate your work into the OpenScPCA project!

JingxuanChen7 commented on analyses/cell-type-wilms-tumor-14/conda_envs/main.yml:
Hi @jashapiro , Thanks so much for offering help! Yes indeed, most of my analysis would use R packages. And I agree that some conda recipes are not up-to-date & too many unnecessary conflicts. So I'd love to transfer to `renv`.

Since I don't have much experience with `renv`, I do have some concerns on installation if transferring to `renv`. When I tried to install R packages with `install.packages("Seurat")` on the virtual computer, I realized that I need to first install some dependency libraries like `sudo apt install libglpk40`. Similar scenarios happened for other packages I used like `infercnv` or `seurat-disk` etc, which all needed manual installation of external libraries. In this case, is it OK to just record package version using `renv::snapshot()`?

jashapiro replied:
That is a great question. We had hoped that the virtual machines would have most of the packages needed, but we obviously didn't get everything on there.

What would probably be best for now is if you could include a list of required system packages in the readme or an installation script with all of the necessary system packages as you discover them. That might look like something as simple as:

```sh
sudo apt install -y \
  libglpk40 \
  libcurl \
...
```

That would be helpful for anyone installing those packages on one of our virtual machines, but what we would probably try to do going forward is make sure that those packages are preinstalled in the future, but that could take some time, and would not apply to any virtual machines that you had already created. 

We would also use that list of packages when we create the Docker image for your package. 

Then you could use `renv::snapshot()` to handle all of packages and versions that are specific to R, which also will make the environment accessible to people working on other systems (assuming they have the required system libraries installed).

JingxuanChen7 replied:
Thanks for the suggestions! I've made following changes:
- Added library dependencies https://github.com/AlexsLemonade/OpenScPCA-analysis/pull/764/commits/487d6e99a1a938f54dc1c352c79925a99207780c
- Removed yaml files to setup conda environments, added snapshot in `renv.lock` instead. Updated related "Usage" session in README. Tested working on my virtual computer. https://github.com/AlexsLemonade/OpenScPCA-analysis/pull/764/commits/8e696213eb9b8d9c712fdf42a7fca0ce9e1489d6

JingxuanChen7 commented on analyses/cell-type-wilms-tumor-14/README.md:
````diff
@@ -0,0 +1,91 @@
+* Create this module structure
+```bash
+./create-analysis-module.py cell-type-wilms-14 --use-r  --use-renv --use-conda --conda-file-only
+```
````
Removed in https://github.com/AlexsLemonade/OpenScPCA-analysis/pull/764/commits/487d6e99a1a938f54dc1c352c79925a99207780c

JingxuanChen7 commented on analyses/cell-type-wilms-tumor-14/notebook-template.Rmd:
Removed `rmd` tempplate in https://github.com/AlexsLemonade/OpenScPCA-analysis/pull/764/commits/487d6e99a1a938f54dc1c352c79925a99207780c

JingxuanChen7 commented:
Hello @jashapiro , thanks for all the help!

> As for the module setup itself, I very much appreciate you setting up a distinct environment for your work with all of your package dependencies included. It seems like your project is mostly based in R? If that is the case, we would in general prefer that you set up your environment with renv, as that allows us to more easily track the versions of specific packages and dependencies. The versions available on conda are often a bit out of date as well, so this allows us to keep up better with current versions of Bioconductor, etc.
> 
> If you would like any help setting that up, please let me know, or if you have a compelling reason that you would prefer to stick with conda for this purpose, we can also consider that for this project.

Thanks so much for the suggestion. I understood that using `renv` is much easier to manage R packages. I have made changes transferring to log versions with `renv` in my latest commit (https://github.com/AlexsLemonade/OpenScPCA-analysis/pull/764/commits/8e696213eb9b8d9c712fdf42a7fca0ce9e1489d6).

> Since I think it makes sense to start with the project as a whole, I did not spend too much time looking at your processing code. I did note, however, that you are performing normalization and dimension reduction within Seurat. This may be required by your downstream analysis, but I will note that we do include default normalized and dimension reduction matrices, and it may be more efficient to maintain the pre-computed values in the SingleCellExperiment object rather than recalculating them. If this is something you would like to pursue, I am happy to answer questions about where those values are located. The objects and processing steps are generally described in the portal documentation at https://scpca.readthedocs.io/en/stable/sce_file_contents.html

Thanks for the information. In my preprocessing code, I modified some parameters in the Seurat workflow (e.g number of features in the feature selection, whether to run `harmony` since I may merge samples in following analysis). In the future, I may want to try different feature selection or clustering algorithms, since cell types are not split out in some samples (preliminary results, not shown in this PR). Therefore, I would like to keep my codes for pre-processing as of now.

I think all the comments have been resolved in my last two commits. Please let me know if I should make any other changes. Thanks again for all the help!

jashapiro reviewed (commented):
Thank you for the updated version @JingxuanChen7! I think it looks good, and I again appreciate the effort you have gone to to provide complete documentation!

I had a few minor suggestions for coding robustness and style, with the most significant and possibly disruptive being about the use of Ensembl gene ids vs. gene symbols. As I said in my comment below, we are generally standardized on the Ensembl IDs for this project, so if you can use those with the Seurat objects, that is likely to yield the most reliable results. I am not sure how downstream processing will handle the results of the deduplication as you are performing it; if the names of the rows are changing in the dedup process, then it will take extra steps to check the gene symbols anyway, and leaving the Ensembl IDs may be more efficient. 

I should also note that it may not be immediately clear what the "correct" thing to do is when there are multiple Ensembl IDs with the same gene symbol. I suspect that the best approach is to sum the counts with the same gene symbol, as the upstream quantification procedure will have "assigned" reads to one or the other Ensembl ID, but I am not certain that will yield the most reliable results, so I would be hesitant to discard the data at this stage. 

If your downstream analysis does rely on gene symbols, it may well be reasonable to discard the features which have `NA` for their gene symbol, as those will never come up (and are more likely to be dubious annotations).

Thanks again for your contribution, and I look forward to seeing your next steps soon (after we get this one in 😉)!

JingxuanChen7 commented on analyses/cell-type-wilms-tumor-14/scripts/utils/00_preprocessing_rds_functions.R:
```diff
@@ -0,0 +1,84 @@
+  # out_rds <- paste0(path_proj,"/",sample,"/",library,"_processed_genesym.rds")
+  db <- read.table(paste0(db_proj,"/",sample,"/",library,"_processed_scdblfinder.tsv"), header = T)
+  
+  ######## convert ensembl IDs to symbols and dedup
+  rownames(rds) <- rds@rowRanges@elementMetadata@listData[["gene_symbol"]]
+  rds <- singleCellTK::dedupRowNames(rds, as.rowData = F, return.list = F)
+  rds <- rds[!is.na(rownames(rds)),]
+  rds <- rds[!duplicated(rownames(rds)),]
+  # readr::write_rds(rds, out_rds, compress = "bz2")
```
> I expect that part of the reason that you are taking this approach is that the marker genes that you will be using later are listed by gene symbol, and if that is the case, I might try converting in the other direction?

I converted to gene symbol because most of reference databases (e.g Human kidney atlas, other public single-cell datasets I used in my test) have gene symbols as row names by default. And for sure, I used gene symbol when curating markers genes. Thus, when testing different methods, it's always more convenient to use gene symbol.

On the other hand, I agree that simply converting to gene symbol and removing duplicates may lose some information, since it's less standard than Ensembl IDs.

To align with your project standard naming, I would keep ensembl IDs as default for this step. I'll look more into reference databases & marker gene list, trying to convert gene symbol to ensembl IDs. Or converting row names to gene symbols for individual sections which don't support ensembl IDs.

> I am a bit confused about why there are still duplicated rows after running singleCellTK::dedupRowNames. Do you have a thought there?

Never mind. Looking back to my scripts, I realize it's redundant. I think kept it just to avoid any error messages.

jashapiro commented on analyses/cell-type-wilms-tumor-14/scripts/utils/00_preprocessing_rds_functions.R:
```diff
@@ -0,0 +1,84 @@
+  # readr::write_rds(rds, out_rds, compress = "bz2")
+  
+  ######## remove doublets
+  rds <- rds[,which(db$class == "singlet")]
```
I don't think I have a very strong preference about the final resolution at this point, as I view this as an empirical question. My main thought is just that preserving some flexibility so that we can look more carefully at the differences later may be beneficial.

JingxuanChen7 commented:
Hi @jashapiro , thank you so much for all the suggestions on coding robustness and style. I really appreciate it!

Regarding the gene symbol vs. Ensembl ID, I keep Ensembl ID at this stage. So I also updated result `.rdsSeurat` files on S3 bucket. Regarding doublets, I added a column in the metadata instead of removing them. Other minor coding style changes have been included in my latest commit https://github.com/AlexsLemonade/OpenScPCA-analysis/pull/764/commits/9be5d74171dfd607f8581ad126e60715023e05ba .

Thanks again for the careful code review. Let me know if anything else needs to be modified!

jashapiro commented on analyses/cell-type-wilms-tumor-14/scripts/utils/00_preprocessing_rds_functions.R:
```diff
@@ -0,0 +1,75 @@
+                          skip_logNorm = TRUE){
+  ######## Normalize, scale, feature selection
+  if (skip_logNorm){
+    obj[["RNA"]]$data <- obj[["RNA"]]$counts
```
I don't think we want to copy the counts, but rather just verify that the `data` slot is present (and of the correct dimensions, perhaps)?
```suggestion
    stopifnot("data layer must be present if skip_logNorm selected" = 
      "data" %in% SeuratObject::Layers(obj[["RNA"]])
    )
```

jashapiro commented on analyses/cell-type-wilms-tumor-14/scripts/utils/00_preprocessing_rds_functions.R:
```diff
@@ -0,0 +1,75 @@
+  rds$class <- db$class
+  
+  ######## create seurat object from the SCE counts matrix
+  if (get_logcounts) {
+    seurat_obj <- SeuratObject::CreateSeuratObject(counts = SingleCellExperiment::logcounts(rds),assay = "RNA",project = library)
+  }else {
+    seurat_obj <- SeuratObject::CreateSeuratObject(counts = SingleCellExperiment::counts(rds),assay = "RNA",project = library)
+  }
```
I think you want to keep the raw counts either way, so maybe:

(Using the SeuratObject function to allow it to check or reformat if desired)

```suggestion
  seurat_obj <- SeuratObject::CreateSeuratObject(counts = SingleCellExperiment::counts(rds),assay = "RNA",project = library)
  
  # add normalized data if requested
  if (get_logcounts) {
    SeuratObject::LayerData(seurat_obj[["RNA"]], "data") <- SingleCellExperiment::logcounts(rds)
  }
```

jashapiro commented on analyses/cell-type-wilms-tumor-14/scripts/utils/00_preprocessing_rds_functions.R:
```diff
@@ -0,0 +1,75 @@
+  # add metadata from SingleCellExperiment to Seurat
+  seurat_obj@misc <- S4Vectors::metadata(rds)
+  
+  skip_logNorm <- get_logcounts
+  seurat_obj <- pre_seuratobj(seurat_obj, nfeatures = 3000, run_harmony = F, reduction = "pca", skip_logNorm = skip_logNorm)
```
Just skipping an extra variable
```suggestion
  seurat_obj <- pre_seuratobj(seurat_obj, nfeatures = 3000, run_harmony = F, reduction = "pca", skip_logNorm = get_logcounts)
```

jashapiro commented on analyses/cell-type-wilms-tumor-14/scripts/utils/00_preprocessing_rds_functions.R:
```diff
@@ -0,0 +1,75 @@
+
+  ######## annotate doublets
+  #rds <- rds[,which(db$class == "singlet")]
+  rds$class <- db$class
```
Just thinking that renaming here might be useful for later clarity
```suggestion
  rds$doublet_class <- db$class
```

jashapiro reviewed (commented):
This looks good to me! I had just a few more minor comments, including keeping the counts and data if using the existing logcounts. But I expect to approve and merge this with the next revision!

jashapiro commented:
Oh, I did have one more comment, which was that you might want to store the `.rdsSeurat` files just in the `scratch/` directory. I am not sure you will need to syncy them to S3, as at this stage they are not really results files so much as reformatted data files. I was able to verify that you are successfully syncing though, which has some value for the future!

JingxuanChen7 commented:
> Oh, I did have one more comment, which was that you might want to store the `.rdsSeurat` files just in the `scratch/` directory. I am not sure you will need to syncy them to S3, as at this stage they are not really results files so much as reformatted data files. I was able to verify that you are successfully syncing though, which has some value for the future!

Hi @jashapiro , thank you for reminding me about the `scratch/` folder! It's more appropriate to put the intermediate files to `scratch/` instead of `results/`. I would re-sync the results in my next PR. 

In addition, I applied changes as suggested in commit https://github.com/AlexsLemonade/OpenScPCA-analysis/pull/764/commits/1731730a028e1a6a044cfb2b460a21deca3ee074 Again, I appreciate all the suggestions, which are really helpful!

jashapiro reviewed (approved):
This looks good to me! Congratulations on your first PR to the project! 

Feel free to proceed to the next step of your analysis. We will soon want to update `.github/workflows/run_cell-type-wilms-tumor-14.yml` to run the analysis on test data as part of our project test procedures, but this can probably wait until you have a bit more done.  When you submit your next step, you may want to include a top-level file of `run_cellytping.sh` or similar that runs the individual scripts in sequence.

As I write that, I realize it may also be helpful down the road to refactor  the `00_preprocessing_rds.R` script (or just write an alternative version) to act on a single library at a time, as this will make it easier to parallelize the analysis down the road. I don't think that needs to be your priority at this stage, but you might consider your downstream scripts taking that form so that they work on one library or sample at a time where possible.
